### DISCUSSION TITLE
Adopting BCP 14 (RFC 2119, RFC 8174) requirement levels in CF

### DISCUSSION THREAD

**Discussion by JonathanGregory** [cf-convention-test/discuss#4] [Comments]
In the discussion of #256, the idea of formally adopting RFC 2119 requirement level key words was raised by myself and, perhaps foolishly 😉, I offered to start the effort if desired. This received immediate support from @larsbarring and @JonathanGregory. So this new issue where I will describe what this is, what changes would need to be adopted, and what I think the steps might be to getting this implemented. Adopting this will likely touch every section of the CF document and impact new and existing proposed changes to CF. [@DocOtak](https://github.com/cf-convention/discuss/issues/258#issue-1854107966)

**Comment by JonathanGregory** [comment]
Things can absolutely take place in a publicly in a branch. I don't think merging in changes from the main branch would be too difficult, there are several options: rebase, merge commits, etc... and I don't think making this change incrementally vs one large changeset change the effort involved for those of us who would be making the actual modifications. Considerations about how external readers might be affected should take precedence in the decisions in my opinion. [@DocOtak](https://github.com/cf-convention/discuss/issues/258#issuecomment-1683667410)

**Reply by JonathanGregory** [reply]
[deleted]

**Comment by larsbarring** [comment]
I did a little bit of playing around with some ideas and put together some python code that paints the selected keywords. Basically it takes the .adoc files located in the parent directory (which is where they are as usual) and adds some asciidoc coloring tags and  saves the results in the same directory as the python code. And if you have asciidoctor installed it creates html in a subdirectory. All is available in [my fork under branch BCP14](https://github.com/larsbarring/cf-conventions/tree/my_bcp-14), in particular in the subdirectory `BCP-14`.

Perhaps this could be a starting point to build on, but I have rather limited experience of github and do not know how to set up a branch that will efficient for collaboration. 

Another thing that I thought of is what words/phrases to work on. BCP14 has a clearly defined list. I think that @DocOtak wanted to restrict the work to only these. I like this because this is a rather extensive undertaking anyway. On the other hand, if we are going to do all this work, I think that it might at least be useful to be aware of closely related words/phrases so we don't miss some important context. In the python code I have these preliminary lists (which have to be checked/updated !!):

```python
BCP14 = [
    "MUST NOT", "SHALL NOT", "SHOULD NOT",
    "MUST", "REQUIRED", "SHALL", "SHOULD",
    "RECOMMENDED", "MAY", "OPTIONAL"
]
EXTENDED_BCP14 = [
    "NOT RECOMMENDED", "RECOMMENDS* NOT","RECOMMENDS*",
    "CAN NOT", "COULD NOT", "CAN", "COULD", 
    "DEPRECATED", "HAVE TO"
```
